### ISSUE TITLE
Bug: varsIgnorePattern also applies to catch arguments

### ISSUE THREAD

**Issue by b13nxx** [eslint/eslint#17540]
### Environment

Node version: v18.17.1
npm version: 9.6.7
Local ESLint version: v8.48.0
Global ESLint version: -
Operating System: Windows

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
'@typescript-eslint/no-unused-vars': [
      'error',
      {
        varsIgnorePattern: '^.',
        caughtErrors: 'all'
      }
    ]
```
</details>

```js
try {
  // ...
} catch (err) { // <== unused variable
  console.error(process.env)
}
```

### What did you expect to happen?

I expected the `caughtErrors: all` setting to catch the `err` variable in the catch.

### What actually happened?

Instead it gave no warning.

### Link to Minimal Reproducible Example

[Playground Example](https://eslint.org/play/#eyJ0ZXh0IjoidHJ5IHtcbiAgLy8gLi4uXG59IGNhdGNoIChlcnIpIHtcbiAgY29uc29sZS5lcnJvcihwcm9jZXNzLmVudilcbn0iLCJvcHRpb25zIjp7ImVudiI6eyJlczYiOnRydWV9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCJ9LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6WyJlcnJvciJdLCJmb3ItZGlyZWN0aW9uIjpbImVycm9yIl0sImdldHRlci1yZXR1cm4iOlsiZXJyb3IiXSwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6WyJlcnJvciJdLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6WyJlcnJvciJdLCJuby1jbGFzcy1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29tcGFyZS1uZWctemVybyI6WyJlcnJvciJdLCJuby1jb25kLWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb25zdC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3RhbnQtY29uZGl0aW9uIjpbImVycm9yIl0sIm5vLWNvbnRyb2wtcmVnZXgiOlsiZXJyb3IiXSwibm8tZGVidWdnZXIiOlsiZXJyb3IiXSwibm8tZGVsZXRlLXZhciI6WyJlcnJvciJdLCJuby1kdXBlLWFyZ3MiOlsiZXJyb3IiXSwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjpbImVycm9yIl0sIm5vLWR1cGUtZWxzZS1pZiI6WyJlcnJvciJdLCJuby1kdXBlLWtleXMiOlsiZXJyb3IiXSwibm8tZHVwbGljYXRlLWNhc2UiOlsiZXJyb3IiXSwibm8tZW1wdHkiOlsiZXJyb3IiXSwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjpbImVycm9yIl0sIm5vLWVtcHR5LXBhdHRlcm4iOlsiZXJyb3IiXSwibm8tZXgtYXNzaWduIjpbImVycm9yIl0sIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6WyJlcnJvciJdLCJuby1leHRyYS1zZW1pIjpbImVycm9yIl0sIm5vLWZhbGx0aHJvdWdoIjpbImVycm9yIl0sIm5vLWZ1bmMtYXNzaWduIjpbImVycm9yIl0sIm5vLWdsb2JhbC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW1wb3J0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8taW52YWxpZC1yZWdleHAiOlsiZXJyb3IiXSwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOlsiZXJyb3IiXSwibm8tbG9zcy1vZi1wcmVjaXNpb24iOlsiZXJyb3IiXSwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjpbImVycm9yIl0sIm5vLW5ldy1zeW1ib2wiOlsiZXJyb3IiXSwibm8tbm9ub2N0YWwtZGVjaW1hbC1lc2NhcGUiOlsiZXJyb3IiXSwibm8tb2JqLWNhbGxzIjpbImVycm9yIl0sIm5vLW9jdGFsIjpbImVycm9yIl0sIm5vLXByb3RvdHlwZS1idWlsdGlucyI6WyJlcnJvciJdLCJuby1yZWRlY2xhcmUiOlsiZXJyb3IiXSwibm8tcmVnZXgtc3BhY2VzIjpbImVycm9yIl0sIm5vLXNlbGYtYXNzaWduIjpbImVycm9yIl0sIm5vLXNldHRlci1yZXR1cm4iOlsiZXJyb3IiXSwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOlsiZXJyb3IiXSwibm8tc3BhcnNlLWFycmF5cyI6WyJlcnJvciJdLCJuby10aGlzLWJlZm9yZS1zdXBlciI6WyJlcnJvciJdLCJuby11bmRlZiI6WyJvZmYiXSwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOlsiZXJyb3IiXSwibm8tdW5yZWFjaGFibGUiOlsiZXJyb3IiXSwibm8tdW5zYWZlLWZpbmFsbHkiOlsiZXJyb3IiXSwibm8tdW5zYWZlLW5lZ2F0aW9uIjpbImVycm9yIl0sIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6WyJlcnJvciJdLCJuby11bnVzZWQtbGFiZWxzIjpbImVycm9yIl0sIm5vLXVudXNlZC12YXJzIjpbImVycm9yIix7InZhcnNJZ25vcmVQYXR0ZXJuIjoiXi4iLCJjYXVnaHRFcnJvcnMiOiJhbGwifV0sIm5vLXVzZWxlc3MtYmFja3JlZmVyZW5jZSI6WyJlcnJvciJdLCJuby11c2VsZXNzLWNhdGNoIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtZXNjYXBlIjpbImVycm9yIl0sIm5vLXdpdGgiOlsiZXJyb3IiXSwicmVxdWlyZS15aWVsZCI6WyJlcnJvciJdLCJ1c2UtaXNuYW4iOlsiZXJyb3IiXSwidmFsaWQtdHlwZW9mIjpbImVycm9yIl19fX0=)

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

Since there are separate ignore settings for `caughtErrors` (**caughtErrorsIgnorePattern**) and `vars` (**varsIgnorePattern**), I think that setting one should not affect the other. In addition to fixing this behavior, we could also add `none` value to `vars` setting, just like `args` and `catchErrors`. Because what the above example is basically trying to do is to **turn off** the behavior of checking the variables.

**Comment by Rec0iL99** [comment]
Hi @b13nxx, thanks for the issue. I think it's a bug that having `varsIgnorePattern: '^.'` is leading to `err` variable in the catch clause being ignored. Pushing it forward for the core team to verify.

**Comment by mdjermanovic** [comment]
The current behavior is indeed confusing.

For `varsIgnorePatern`, I think it would make sense to apply either to all variables or only to those not covered by other ignore pattern options.

Currently, `varsIgnorePatern` applies to all variables except for parameters, because they are covered by `argsIgnorePattern`. By the same logic, it shouldn't apply to `catch ()` variables as well, because they are covered by `caughtErrorsIgnorePattern`.

I believe this was for backwards compatibility reasons. `varsIgnorePattern` and `argsIgnorePattern` were added at the same time (https://github.com/eslint/eslint/pull/3433), while `caughtErrorsIgnorePattern` was added later (https://github.com/eslint/eslint/pull/5607). At that point (and now as well), updating `varsIgnorePattern` to no longer apply to those variables would be a breaking change.

I'd support this change for v9. Curious what other team members think about this.

**Comment by nzakas** [comment]
> I believe this was for backwards compatibility reasons. varsIgnorePattern and argsIgnorePattern were added at the same time (https://github.com/eslint/eslint/pull/3433), while caughtErrorsIgnorePattern was added later (https://github.com/eslint/eslint/pull/5607). At that point (and now as well), updating varsIgnorePattern to no longer apply to those variables would be a breaking change.

I think this change makes sense for v9.
